packages/react/src/ReactAct.js

```diff
@@ -91,30 +91,23 @@ export function act<T>(callback: () => T | Thenable<T>): Thenable<T> {
 
       if (__DEV__) {
         if (!didWarnNoAwaitAct && typeof Promise !== 'undefined') {
-          const awaitDetector = () => {
-            if (!wasAwaited) {
-              didWarnNoAwaitAct = true;
-              console.error(
-                'You called act(async () => ...) without await. ' +
-                  'This could lead to unexpected testing behaviour, ' +
-                  'interleaving multiple act calls and mixing their ' +
-                  'scopes. ' +
-                  'You should - await act(async () => ...);',
-              );
-            }
-          };
-          if (result instanceof Promise && 'finally' in result) {
-            // Using finally here is more compatible with native promises
-            // and promise libraries like core.js
-            result
-              .finally(() => {})
-              .catch(() => {})
-              .then(awaitDetector);
-          } else {
-            Promise.resolve()
-              .then(() => {})
-              .then(awaitDetector);
-          }
+          Promise.resolve(result)
+            .then(
+              () => {},
+              () => {},
```

Comment:
Whats the reason for this .then block with two anonymous functions in it? In my testing we can just remove this .then. The error is fixed by putting the result inside the Promise.resolve.

Reply:
I think that it is commendable to try detect common errors and emit a warning, but maybe a project of this scale should not push the limits of the language and if something cannot be done in a reliable way, it probably should not be attempted.

Do the language specifications enforce the `then()` order? Or is it simply Node.js' implementation?

Look at this: https://github.com/callstack/react-native-testing-library/issues/379

Reply:
Yes, the specification forces the ordering.

Reply:
Ok, I found it: 27.2.5.4.1.9a. Append fulfillReaction as the last element of the [List](https://tc39.es/ecma262/#sec-list-and-record-specification-type) that is promise.[[PromiseFulfillReactions]].

This means that every environment that has multiple Promise implementations is potentially not conforming to the specifications.

Reply:
There are no such environments. A userland Promise library is irrelevant.

Reply:
React Native is one such environment. If you transpile the code and use a non-standard Promise library, and especially if there are two different implementations - one of them using `asap`  - such as in this issue - then this is possible.

Reply:
> and use a non-standard Promise library

that's not RN, that's you using a Promise library.

If you're a) allowing more than one Promise library to be in your dep graph, or b) using a Promise library that doesn't follow the spec, then both of those are problems you need to fix.

Reply:
No, it is React Native, it replaces the standard Promise constructor with one that relies on `asap`.

If you transpile this code, it can potentially fail - even if every Promise shim follows the spec - it simply takes having two of them - something that cannot always be easily controlled.

Which is the reason I suggested to not push the language to the limit in a project such as React - as it will get used and abused in different environments - and it will be very often transpiled.

Reply:
Sounds like an issue should be filed on RN then - it's not scalable for everything else in the ecosystem to work around specific bugs.